subsys: fs: nvs: add possibility to check whether record read was the last in history

baw-serafin opened the pull request:
New nvs funtion nvs_read_hist_more
------------------------------------------------
The new API allows to better differantiate between an error and no
more date available when looping over all entries.

Also adapt the sample to use new API

Commit:
nvs: Add API for nvs_read_hist_more

The current APi makes it hard to loop over every
entriy of a given ID and differentiate between
error and no more data.

This is just the API the funtionnality is not yet included

Signed-off-by: serafin <serafin@bytesatwork.ch>

Commit:
nvs: implement nvs_read_hist_more

Signed-off-by: serafin <serafin@bytesatwork.ch>

Commit:
nvs sample: Use nvs_read_hist_more to loop over data

Signed-off-by: serafin <serafin@bytesatwork.ch>

baw-serafin commented on subsys/fs/nvs/nvs.c:
```diff
@@ -831,12 +832,28 @@ ssize_t nvs_read_hist(struct nvs_fs *fs, u16_t id, void *data, size_t len,
+	ret = wlk_ate.len;
+
+	if (more != NULL) {
+		(void) _nvs_prev_ate(fs, &wlk_addr, &wlk_ate); //save to ignore?
```
Obviously the comment needs to go. Any Idea how best to handle an error on '_nvs_prev_ate'?

nvlsianpu replied:
This should be in the loop (as above) as you want to check existence of record of the known ID.
error shouldn't happened as ate was read in the last.

Error shouldn't happen - but pleas serve it as it is served above.

nvlsianpu replied:
@baw-serafin 
>This should be in the loop ...

Was you intention to check existence of older entry of the sane ID, or any older entry?

baw-serafin replied:
Should be fixed now also rebased to curent master to include the style fixes

nvlsianpu commented:
@Laczen If you can look at this it would be great.

Laczen commented:
Hi @baw-serafin and @nvlsianpu, I don't really see a reason to extend the API. To distinguish between an item not found and a error you should look at the rc return value. In case there is no more entry the return value should be -ENOENT.

Laczen commented:
@nvlsianpu, the proposed change does not improve execution time. The ``more`` variable is only reflecting if a read was succesful, not that there is an extra value. Checking that the read was succesful could also be done by looking at the return value: as long as the return value is not -ENOENT (or zero) you keep the loop running, as soon as you reach -ENOENT you jump out of the counting loop, if a different return non-zero value is seen you can do different error handling.

There is even another disadvantage in using the ``more`` variable, you always need to initialize it to zero for a different entry (or a new loop for the same variable: because a variable could be deleted or a gc could be done). So the meaning of the ``more`` variable is only valid during a started loop.

nvlsianpu commented:
@baw-serafin Can you answer to @Laczen doubts. Think that what he suggest works for you as well. Or maybe ti would be better to ju to have an iterator which will run a handler for every encounters entry of an ID? (similar `walk` as fcb has).

baw-serafin commented:
> @baw-serafin Can you answer to @Laczen doubts. Think that what he suggest works for you as well. Or maybe ti would be better for your case to have an iterator which will run a handler for every encounters entry of an ID? (similar `walk` as fcb has).

@nvlsianpu Sorry was away last week will look onto this in the next few days

Commit:
fixup! nvs: implement nvs_read_hist_more

extract looping over ate into inline funtion, as we need to loop
as well in case *more != null

This also makes the error handling easier and the whole function easier to
read.

Laczen commented:
Hi @baw-serafin, thanks for your efforts on improving the nvs subsystem.

As I mentioned before I think there are some problems with it.

First of all I think the improvement is very limited as nvs_read_hist already has the ability to distinguish between an error and no more data available, the return value is -ENOENT if there is no more data available.

A second problem I see with it is that it the routine nvs_read_hist_more is subject to errors: the history data of an entry can be changed by every write that is done to nvs, so this makes the more variable limited to the case of a loop where there are no writes.

The optimisation is also limited to a very rare case, only the last nvs_read_hist() is avoided, but all previous nvs_read_hist_more are doing more flash accesses. I think this even means the performance is not improved because all reads that are performed during the last nvs_read_hist() that returns -ENOENT are performed before in each nvs_read_hist_more() in finding if there is a next entry.

Maybe I am overlooking something so please comment. Anyhow, thanks again for the efforts.

baw-serafin commented on subsys/fs/nvs/nvs.c:
```diff
@@ -790,14 +790,38 @@ int nvs_delete(struct nvs_fs *fs, u16_t id)
+		}
+	} while (*wlk_addr != fs->ate_wra);
+
+	if (*wlk_addr == fs->ate_wra) {
```
Not shure this holds if found object is the last one. @Laczen ?

Laczen commented:
Hi @baw-serafin, thanks for your comments. Now that I understand your use case we can focus on solutions.

Now that you know that there is something like -ENOENT (Error No entry) could you not change the method for the overlaying protocol: Read the data you need as well as the next data to see if there is one and send that it was the last if there is no more next data (-ENOENT) ? This could further be optimized by keeping the data for the next transmission. In that case you would avoid reading twice.

Also don't forget you will need to handle the -ENOENT anyhow because it is not guaranteed that more is still valid. If you handle it in one case your solution should become cleaner and simpler.

Please also provide you use case as a sample or use case, other people will be trying to do exactly the same as you are, an example of how someone solved this could be very interesting.

Regarding if an error code is expected behavior or not: nvs is not smart, it is asked to supply the nth data from a certain id, nvs answers: this data is not found.

baw-serafin commented:
> Now that you know that there is something like -ENOENT (Error No entry) could you not change the method for the overlaying protocol: Read the data you need as well as the next data to see if there is one and send that it was the last if there is no more next data (-ENOENT) ? This could further be optimized by keeping the data for the next transmission. In that case you would avoid reading twice.

Yes this would be possible, but reading twice would be very inefficient (we will have 1000s of measurments) so I would indeed need to cache and that is a bit of a hassle. It's not that complicated but I thinkt the only reasonable usecase for `nvs_read_hist`, is to loop over all the entries so the API should make that as easy as possible. That's the reason I made a PR. Making my protocol work, will be no problem. I can even change the protocol, If I have to. but I believe looping should be easier as it is atm.
 
> Also don't forget you will need to handle the -ENOENT anyhow because it is not guaranteed that more is still valid. If you handle it in one case your solution should become cleaner and simpler.

Yes this "TOCTTOU" is a bit annoying but for the simple usecase of looping it will seldom be a problem so I think one can get away with treating ENOENT like any other error in this case.

> Please also provide you use case as a sample or use case, other people will be trying to do exactly the same as you are, an example of how someone solved this could be very interesting.

My use case is not that different from the one in the offical sample I just want all the entries, only difference is that I am not in a loop but polled.

> Regarding if an error code is expected behaviour or not: nvs is not smart, it is asked to supply the nth data from a certain id, nvs answers: this data is not found.

Yes that why I believe looping is the only common use case so it should be easy ;)

Laczen commented:
> > Now that you know that there is something like -ENOENT (Error No entry) could you not change the method for the overlaying protocol: Read the data you need as well as the next data to see if there is one and send that it was the last if there is no more next data (-ENOENT) ? This could further be optimized by keeping the data for the next transmission. In that case you would avoid reading twice.
> 
> Yes this would be possible, but reading twice would be very inefficient (we will have 1000s of measurments) so I would indeed need to cache and that is a bit of a hassle. It's not that complicated but I thinkt the only reasonable usecase for `nvs_read_hist`, is to loop over all the entries so the API should make that as easy as possible. That's the reason I made a PR. Making my protocol work, will be no problem. I can even change the protocol, If I have to. but I believe looping should be easier as it is atm.
> 
If you can make changes to the protocol it is even easier, just ask for the next data, if there is no more return a "no more data available" message. No need to double read or keeping track of previous data. It can't be simpler than this. Only one extra communication at the end but no need to transmit that there is more data available at each data request.

> > Also don't forget you will need to handle the -ENOENT anyhow because it is not guaranteed that more is still valid. If you handle it in one case your solution should become cleaner and simpler.
> 
> Yes this "TOCTTOU" is a bit annoying but for the simple usecase of looping it will seldom be a problem so I think one can get away with treating ENOENT like any other error in this case.
>
That you might get away with it is not sufficient for general use. If the more variable says there is more data then there should be more data.
> > Please also provide you use case as a sample or use case, other people will be trying to do exactly the same as you are, an example of how someone solved this could be very interesting.
> 
> My use case is not that different from the one in the offical sample I just want all the entries, only difference is that I am not in a loop but polled.
> 
> > Regarding if an error code is expected behaviour or not: nvs is not smart, it is asked to supply the nth data from a certain id, nvs answers: this data is not found.
> 
> Yes that why I believe looping is the only common use case so it should be easy ;)
It is easy, nvs_read() returns 0 if the data is found (and provides the data) and returns an error code if the data is not found (meaning also that there is no data provided) or in case of hardware error.

I do agree that the sample could use some improvement and that distinguishing between a hardware error or no more data available could be made.

baw-serafin commented:
@Laczen Fixing my problem won't be an issue. I just believe you always wan't to have information whether there is more data if you are reading the history. There is no use case for reading the 7th-oldest entry out of the blue. Thats why I asked @nvlsianpu if he's interested in a solutions like this (his name came up in the git log).  But if you are not interested I will solve my problem another way.

In the case the nvs subsys actually deleted, the entry before one could read it, it is an actual error: I have expected data, but it is no more available, so I will probably log an error as if I would do if the read did fail. In the case where I have to read an additional time to check if there is data, the error is not an error at all and I would not want to log it.

>  no need to transmit that there is more data available at each data request.

This one bit I can spare ;)

nvlsianpu commented:
@Laczen I think your opinion is still needed - would you want to accept this patch or refuse it?

Laczen reviewed (changes requested):
Hi @baw-serafin,

Thanks again for proposing a PR for the nvs subsystem.

As it is I would refuse the PR, the main consideration I have is the following: the definition of the more variable is not inline with the volatile nature of the history of items. As the history of items is volatile and changes whenever a garbage collection is called it is not possible to define if there are more variables available for reading. The definition of the more variable would also imply that nvs has some knowledge about what is stored, at the moment there is no knowledge at all and I would like to keep it this way.

I think something like what you propose could be done but with quite some work as a layer on top of nvs, but not in nvs itself, which would then keep track of history items and reset the more variables if required.

I hope you understand my position and would like to thank you again for the proposal.

nvlsianpu commented:
I'm agree with @Laczen.  He is right regard that historical data existence are not  guaranteed as NVS is not a logos storage. Even if you just check that there is a historical record, it is possible that it expires just before you read it.

@baw-serafin  Excuse my misunderstanding while you ask me for opinion about this functionality.

baw-serafin commented:
@Laczen No problem. I am still not convinced though. I think most application won't need that guaranty , that the data will still be ready the time they read it. If so it is trivial to implement in the application just synchronizing the code reading and writing to the storage. But if you don't want the patches I will solve the problem differently on the application level.

You can also salvage only parts of it. I think the code of the function was really hard to read and understand and the refactoring in two loops made it more readable.

@andrzej-kaczmarek Thx for your support anyway
